pkg/modifycontroller/modify_volume.go

```diff
@@ -53,59 +52,77 @@ func (ctrl *modifyController) modify(pvc *v1.PersistentVolumeClaim, pv *v1.Persi
+	// The in-memory uncertain state will be updated after the next ControllerModifyVolume call.
+	return ctrl.validateVACAndModifyVolumeWithTarget(pvc, pv)
+}
+
+func (ctrl *modifyController) getTargetVAC(pvc *v1.PersistentVolumeClaim, vacName string) (*storagev1.VolumeAttributesClass, error) {
+	vac, err := ctrl.vacLister.Get(vacName)
+	// Check if pvcSpecVac is valid and exist
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			ctrl.eventRecorder.Eventf(pvc, v1.EventTypeWarning, util.VolumeModifyFailed, "VAC %q does not exist.", vacName)
+		}
+		return nil, fmt.Errorf("get VAC with vac name %s in VACInformer cache failed: %w", vacName, err)
+	}
+	return vac, nil
 }
 
 // func validateVACAndModifyVolumeWithTarget validate the VAC. The function sets pvc.Status.ModifyVolumeStatus
 // to Pending if VAC does not exist and proceeds to trigger ModifyVolume if VAC exists
 func (ctrl *modifyController) validateVACAndModifyVolumeWithTarget(
 	pvc *v1.PersistentVolumeClaim,
 	pv *v1.PersistentVolume) (*v1.PersistentVolumeClaim, *v1.PersistentVolume, error, bool) {
-	// The controller only triggers ModifyVolume if pvcSpecVacName is not nil nor empty
-	pvcSpecVacName := pvc.Spec.VolumeAttributesClassName
-	// Check if pvcSpecVac is valid and exist
-	vac, err := ctrl.vacLister.Get(*pvcSpecVacName)
-	if err == nil {
-		// Mark pvc.Status.ModifyVolumeStatus as in progress
-		pvc, err = ctrl.markControllerModifyVolumeStatus(pvc, v1.PersistentVolumeClaimModifyVolumeInProgress, nil)
-		if err != nil {
-			return pvc, pv, err, false
-		}
-		// Record an event to indicate that external resizer is modifying this volume.
-		ctrl.eventRecorder.Event(pvc, v1.EventTypeNormal, util.VolumeModify,
-			fmt.Sprintf("external resizer is modifying volume %s with vac %s", pvc.Name, *pvcSpecVacName))
-		return ctrl.controllerModifyVolumeWithTarget(pvc, pv, vac, pvcSpecVacName)
-	} else {
-		if apierrors.IsNotFound(err) {
-			ctrl.eventRecorder.Eventf(pvc, v1.EventTypeWarning, util.VolumeModifyFailed, "VAC "+*pvcSpecVacName+" does not exist.")
-		}
-		klog.Errorf("Get VAC with vac name %s in VACInformer cache failed: %v", *pvcSpecVacName, err)
+
+	vac, err := ctrl.getTargetVAC(pvc, *pvc.Spec.VolumeAttributesClassName)
+	if err != nil {
 		// Mark pvc.Status.ModifyVolumeStatus as pending
 		pvc, err = ctrl.markControllerModifyVolumeStatus(pvc, v1.PersistentVolumeClaimModifyVolumePending, nil)
 		return pvc, pv, err, false
 	}
+
+	// Mark pvc.Status.ModifyVolumeStatus as in progress
+	pvc, err = ctrl.markControllerModifyVolumeStatus(pvc, v1.PersistentVolumeClaimModifyVolumeInProgress, nil)
+	if err != nil {
+		return pvc, pv, err, false
+	}
+	// Record an event to indicate that external resizer is modifying this volume.
+	ctrl.eventRecorder.Event(pvc, v1.EventTypeNormal, util.VolumeModify,
+		fmt.Sprintf("external resizer is modifying volume %s with vac %s", pvc.Name, vac.Name))
+	return ctrl.controllerModifyVolumeWithTarget(pvc, pv, vac)
```

Comment:
Can you fix the implementation in `controllerModifyVolumeWithTarget`, so as we aren't updating PVC twice for infeasible errors? (basically same as comment I left on 512).